### PR TITLE
fix: corrects the navigation issue if the student does not pass the entrance exam

### LINKED ISSUE #1415: [Bug] Students have access to course content even if the Entrance Exam wasn't finished
### STR:

1. [Enable feature ENTRANCE_EXAMS](https://docs.openedx.org/projects/edx-platform/en/latest/references/featuretoggles.html#featuretoggle-FEATURES['ENTRANCE_EXAMS']) 
2. Set Entrance exam for course in schedule and details
image
3. As a student, enroll to the course

### Actual Result

Students have access to the full course even before the Entrance Exam is finished
image

### Expected Result

Students don't have access to course content before passing with necessary score for entrance exam
image

 

### Notes:
Result of investigation:

The problem is Learning MFE used sections from course_outline. This data was fetched from `/edx-platform/openedx/core/djangoapps/content/learning_sequences/api/outlines.py`, `get_user_course_outline_details`. This function doesn`t check if is there an entrance exam and if the user passes the exam.

An example of how to check if the chapter is an entrance exam and did user passed it can be found at `edx-platform/lms/djangoapps/courseware/views/index.py`,` _add_entrance_exam_to_context` function

### PR DESCRIPTION
**Pull request by imasdekar**
- adds checks while rendering the next button text for status of the entrance exam.

closes #1415

Signed-off by: Ishan Masdekar <imasdekar@disroot.org>

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @imasdekar!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/committers-frontend-app-learning`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Comment by imasdekar** [comment]
@mphilbrick211 Can you please get the checks run and get this reviewed?

**Comment by GlugovGrGlib** [comment]
@imasdekar I believe we need to wait another week until @bradenmacdonald and @farhaanbukhsh take on maintainer responsibilities for this repo, after this it can be reviewed and scheduled for merge

**Comment by itsjeyd** [comment]
Hi @imasdekar, thank you for this contribution!

**Comment by itsjeyd** [comment]
OSPR management note: This PR includes [user-facing changes](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F). However, it fixes an existing bug (#1415) so there should be no need to send it through product review.

CC @jmakowski1123

File: src/course-home/progress-tab/certificate-status/CertificateStatus.jsx [hunk 1]
```diff
@@ -42,13 +46,16 @@ const CertificateStatus = ({ intl }) => {
     certificateAvailableDate,
   } = certificateData || {};
 
+  const entranceExamPassed = entranceExamData ? entranceExamData.entranceExamPassed : null;
```
Comment (bradenmacdonald) [inline src/course-home/progress-tab/certificate-status/CertificateStatus.jsx]: Nit - you can write this more concisely as:
```suggestion
  const entranceExamPassed = entranceExamData?.entranceExamPassed ?? null;
```

File: src/courseware/course/course-exit/utils.js [hunk 2]
```diff
@@ -82,8 +89,9 @@ function GetCourseExitNavigation(courseId, intl) {
     userHasPassingGrade,
     courseExitPageIsActive,
     canViewCertificate,
+    entranceExamPassed,
   );
-  const exitActive = exitMode !== COURSE_EXIT_MODES.disabled;
+  const exitActive = ((exitMode !== COURSE_EXIT_MODES.disabled) && (exitMode !== COURSE_EXIT_MODES.entranceExamFail));
```
Comment (bradenmacdonald) [inline src/courseware/course/course-exit/utils.js:100]: Can you help me understand this? What exactly does `exitActive` mean and what are the various exit modes? Just want to make sure this logic is correct, but I'm not clear on what's even going on here.

I'd love if you could add a few JSDoc comments to this file as part of this PR to explain what an "exit mode" is and what each of the `COURSE_EXIT_MODES` mean. And why `exitActive` should be `false` when the exit mode is `entranceExamFail` ??

Comment (imasdekar) [reply src/courseware/course/course-exit/utils.js:100]: exitActive is used to enable/disable the exit mode i.e. the next buttons. 
Available COURSE_EXIT_MODES: disabled, celebration, nonPassing, inProgress, entranceExamFail
So if the user has failed the entrance exam, access to further course sections should be disabled

Comment (imasdekar) [reply src/courseware/course/course-exit/utils.js:100]: I've added some comments in the code as per your suggestion. Let me know if more explanation is needed in the comments and I will add accordingly.

File: src/course-home/progress-tab/certificate-status/CertificateStatus.jsx (hunk 1 above)
Comment (imasdekar) [reply src/course-home/progress-tab/certificate-status/CertificateStatus.jsx]: Yes, thanks for the suggestion. Have made the changes.

**Review by bradenmacdonald — APPROVED** [review]
Thanks for the fix. I tested it and confirmed the issue is fixed. Although there is still obviously a major bug on the backend - it should be refusing to provide the rest of the course to the user who hasn't passed the exam, regardless of what logic we have on the frontend.

**Review by bradenmacdonald — CHANGES_REQUESTED** [review]
@imasdekar Actually could you please add a test case for this too? That will fix the coverage problem. Sorry, I just noticed now there isn't a new test.

**Comment by imasdekar** [comment]
@bradenmacdonald before implementing this fix on the frontend I dis check the backend and found that the entrance data is sent to the frontend and the respective data has a field which states that the user has failed the entrance exam. Hence I decided to utilise the same and implement a fix on the frontend.

**Comment by imasdekar** [comment]
I will add a test case and update this PR.

**Comment by imasdekar** [comment]
@bradenmacdonald If req. is also that further course data not be sent from the backend at all if the entrance exam status is failed then what exactly should be the response in such a case? I can work on this backend fix as well.

**Comment by bradenmacdonald** [comment]
Great! Replied on #1415.

**Comment by imasdekar** [comment]
@bradenmacdonald  I've added tests as required.

**Comment by bradenmacdonald** [comment]
Thanks @imasdekar ! Unfortunately the test is not passing on CI - can you please take a quick look?

**Comment by imasdekar** [comment]
@bradenmacdonald apologies, it should work now. I missed to push a small update.

**Review by bradenmacdonald — APPROVED** [review]

**Pull request merged**